ui-v2/app/components/phrase-editor.js

```diff
@@ -1,44 +1,67 @@
 import Component from '@ember/component';
 import { get, set } from '@ember/object';
+import { inject as service } from '@ember/service';
 
 export default Component.extend({
+  dom: service('dom'),
   classNames: ['phrase-editor'],
   item: '',
-  remove: function(index, e) {
-    this.items.removeAt(index, 1);
-    this.onchange(e);
+  didInsertElement: function() {
+    this._super(...arguments);
+    // TODO: ref
+    this.input = get(this, 'dom').element('input', this.element);
   },
-  add: function(e) {
-    const value = get(this, 'item').trim();
-    if (value !== '') {
-      set(this, 'item', '');
-      const currentItems = get(this, 'items') || [];
-      const items = new Set(currentItems).add(value);
-      if (items.size > currentItems.length) {
-        set(this, 'items', [...items]);
-        this.onchange(e);
+  onchange: function(e) {},
+  search: function(e) {
+    // TODO: Temporarily continue supporting `searchable`
+    let searchable = get(this, 'searchable');
+    if (searchable) {
+      if (!Array.isArray(searchable)) {
+        searchable = [searchable];
       }
+      searchable.forEach(item => {
+        item.search(get(this, 'value'));
+      });
     }
+    this.onchange(e);
   },
-  onkeydown: function(e) {
-    switch (e.keyCode) {
-      case 8:
-        if (e.target.value == '' && this.items.length > 0) {
-          this.remove(this.items.length - 1);
+  oninput: function(e) {},
+  onkeydown: function(e) {},
+  actions: {
+    keydown: function(e) {
+      switch (e.keyCode) {
+        case 8: // backspace
+          if (e.target.value == '' && get(this, 'value').length > 0) {
+            this.actions.remove.bind(this)(get(this, 'value').length - 1);
+          }
+          break;
+        case 27: // escape
+          set(this, 'value', []);
+          this.search({ target: this });
+          break;
+      }
+      this.onkeydown({ target: this });
+    },
+    input: function(e) {
+      set(this, 'item', e.target.value);
+      this.oninput({ target: this });
+    },
+    remove: function(index, e) {
+      get(this, 'value').removeAt(index, 1);
+      this.search({ target: this });
+      this.input.focus();
+    },
+    add: function(e) {
+      const item = get(this, 'item').trim();
+      if (item !== '') {
+        set(this, 'item', '');
+        const currentItems = get(this, 'value') || [];
```

Comment:
I get the desire to align with DOM APIs here, but I personally don't like that the singular value is being used to represent a collection.

Reply:
Yeah agree here, I almost `split` it by `\n` to make it a string and then `join`ed it back up again on the other side. It was also `detail.value` at one point using `CustomEvent`. I think the thing that makes this 'ok' for me is the fact that as I'm using ember/javascript there is nothing to be gained in splitting/joining things apart from absolute parity with DOM APIs, I'm not sure there is anything practical to be gained form it (at least the `target.value` is recognisable and and therefore clear what is going on)

I'll have another look today though.